lightning/src/ln/channelmanager.rs

```diff
@@ -466,6 +481,42 @@ pub struct ChannelDetails {
 	pub is_live: bool,
 }
 
+/// If a payment fails to send, it can be in one of several states. This enum is returned as the
+/// Err() type describing which state the payment is in, see the description of individual enum
+/// states for more.
+#[derive(Debug)]
+pub enum PaymentSendFailure {
+	/// A parameter which was passed to send_payment was invalid, preventing us from attempting to
+	/// send the payment at all. No channel state has been changed or messages sent to peers, and
+	/// once you've changed the parameter at error, you can freely retry the payment in full.
+	ParameterError(APIError),
+	/// A parameter in a single path which was passed to send_payment was invalid, preventing us
+	/// from attempting to send the payment at all. No channel state has been changed or messages
+	/// sent to peers, and once you've changed the parameter at error, you can freely retry the
+	/// payment in full.
+	///
+	/// The results here are ordered the same as the paths in the route object which was passed to
+	/// send_payment.
+	PathParameterError(Vec<Result<(), APIError>>),
+	/// All paths which were attempted failed to send, with no channel state change taking place.
+	/// You can freely retry the payment in full (though you probably want to do so over different
+	/// paths than the ones selected).
+	AllFailedRetrySafe(Vec<APIError>),
+	/// Some paths which were attempted failed to send, though possibly not all. At least some
+	/// paths have irrevocably committed to the HTLC and retrying the payment in full would result
+	/// in over-/re-payment.
+	///
+	/// The results here are ordered the same as the paths in the route object which was passed to
+	/// send_payment, and any Errs which are not APIError::MonitorUpdateFailed can be safely
+	/// retried (though there is currently no API with which to do so).
```

Comment:
I'm not 100% sure it's reasonable to expect the user to have access to the original route object so they know which paths are failed-but-safe-to-retry, but OK to leave it as is for now. 

In this situation, it could be ergonomic for each Ok(()) or MonitorUpdateFailed() message to contain the amount that "successfully" sent, so the user could immediately make a new make-up payment to get to the originally desired amount.

UX seems acceptable to me for this PR but just brainstorming for the future!

Reply:
Hmm, good point. Sadly in this patch we started clone()ing bits out of the Route anyway, so there's no harm in just taking it by reference, which makes this a bunch better. I added a commit to do so.